Simplify ResourcePublisher

drewnoakes opened the pull request:
We publish updates to resources via `IAsyncEnumerable<ResourceChange>`. The previous code defined a custom enumerable/enumerator for this, but it's possible to use a [generator function](https://learn.microsoft.com/en-us/dotnet/csharp/asynchronous-programming/async-return-types#async-streams-with-iasyncenumerablet) instead and reduce the amount of code a bit.
 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/dotnet/aspire/pull/1346)

drewnoakes committed:
Simplify ResourcePublisher

We publish updates to resources via `IAsyncEnumerable<ResourceChange>`. The previous code defined a custom enumerable/enumerator for this, but it's possible to use a generator function instead and reduce the amount of code a bit.

JamesNK reviewed (approved):
Changes look fine.

Is there test coverage of this area? If not, it would be good to have some to verify it works the same before/after.

drewnoakes commented:
/azp run

drewnoakes committed:
Rename ResourceChangeType.Deleted -> Delete

It's more consistent with the other enum member, named Upsert (which should otherwise be Upserted).

drewnoakes committed:
Add Async suffix to async method

drewnoakes committed:
Add unit tests for ResourcePublisher

drewnoakes commented:
@JamesNK I added `ResourcePublisherTests`.
